### PR TITLE
[sender] reset buffers on forks and reset the companion thread if dead or nil

### PR DESCRIPTION
**Pull request by remeh**
We want to reset the `MessageBuffer` if we appear to be in a recent fork, since the contained messages belong to the parent process.

On top of that, when using the original multi-threaded Sender, this PR is re-spawning the companion thread if it seems to be dead for whatever reasons. It should automatically help when someone is using the multi-threaded mode in an app or framework using forks.

For some reasons, I had to remove the use of `UDPSocket.new.tap` since it was causing the interpreter to crash as soon as there was re-creation of a thread in fork childs.

This is a follow-up of and is replacing #199.

### CODE REVIEW

File: lib/datadog/statsd/udp_connection.rb [hunk 1]
```diff
@@ -19,18 +19,21 @@ def initialize(host, port, **kwargs)
...
 
       def connect
-        UDPSocket.new.tap do |socket|
-          socket.connect(host, port)
-        end
+        @socket.close unless @socket == nil
+
+        @socket = UDPSocket.new
+        @socket.connect(host, port)
       end
```
Comment (ivoanjo) [inline lib/datadog/statsd/udp_connection.rb:33]: This change is ... weird in the context of `Connection#socket`. That method is 

```ruby
      def socket
        @socket ||= connect
      end
```

BUT because `connect` now manages the assignment, that `||=` will never actually do anything.

It looks to me that currently there's not even a code path where it ever gets executed for `UDPConnection`, but since `@socket.connect` does not return the `@socket`, if somehow `@socket == nil`  and `#socket` gets called, the `||=` will assign the result of `@socket.connect` to the `@socket`.

TL;DR, to avoid the complexity, I would suggest not assigning to the `@socket` directly OR to remove the `||=` from `#socket` and make it just an `attr_reader`.

**Review by ivoanjo — COMMENTED** [review]
Left a few notes! One general note is, since our customers do read through these PRs, I suggest renaming the PR to something clearer, such as "Fix metrics reporting in applications using fork" or something similar.

I also recommend clarifying in what situation `UDPSocket.new.tap` was causing the VM to crash, since this seems quite relevant to users on older versions that does use that code (should they upgrade ASAP to avoid this crash? Does it only get triggered along with other changes in this PR?).

File: lib/datadog/statsd/sender.rb [hunk 2]
```diff
@@ -11,9 +18,9 @@ def initialize(message_buffer)
 
       def flush(sync: false)
         # don't try to flush if there is no message_queue instantiated
-        return unless message_queue
+        return unless @message_queue
 
-        message_queue.push(:flush)
+        @message_queue.push(:flush)
```
[the remark this answers is not in the document]
Comment (remeh) [reply lib/datadog/statsd/sender.rb:23]: Thanks for the feedback, I'm re-introducing the `attr_reader`.

File: lib/datadog/statsd/sender.rb [hunk 3]
```diff
@@ -64,25 +79,20 @@ def stop(join_worker: true)
...
 
-      attr_reader :message_buffer
-
-      attr_reader :message_queue
-      attr_reader :sender_thread
-
       if CLOSEABLE_QUEUES
         def send_loop
-          until (message = message_queue.pop).nil? && message_queue.closed?
+          until !@sender_thread.alive? || ((message = @message_queue.pop).nil? && @message_queue.closed?)
```
[the remark this answers is not in the document]
Comment (remeh) [reply lib/datadog/statsd/sender.rb]: True, changing to `Thread.current` for this call.

[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/statsd/sender.rb]: Hmm, on hindsight my question wasn't very straightforward, what I wanted to ask was -- if a thread is checking if it itself is alive, when would `alive?` ever return `false`? It seems like asking the classic "are you asleep?" and expecting a honest yes ;)

File: lib/datadog/statsd/sender.rb [hunk 4]
```diff
@@ -23,20 +30,28 @@ def rendez_vous
...
+        raise ArgumentError, 'Start sender first' unless @message_queue
 
-        message_queue << message
+        # if the thread does not exist, we are probably running in a forked process,
+        # empty the message queue (these messages belong to parent process) and spawn
+        # a new companion thread.
+        if !@sender_thread.alive?
+          @message_queue = nil
+          start
+        end
```
[the remark this answers is not in the document]
Comment (remeh) [reply lib/datadog/statsd/sender.rb]: *  (1) (3) 👍 
* About (2), we would naively try to re-create a thread. I could maybe track how many times we are creating a new thread per second, and emit a warning if it is quite high, but not sure if that's worth. Except from emitting a log message, I don't think there is much we can do since it could be normal in some environments with a lot of forks.

[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/statsd/sender.rb]: Yeah, I think a log message would be nice. The throttling mechanism is an interesting idea, but perhaps not worth the complexity without a clearer use case than "well I what if the thread dies for unknown reasons".

**Comment by ivoanjo** [comment]
One extra thing that occurred to me is that some of the changes to the `Sender` classes are probably not thread-safe.

Since the intention seems to be that one `Sender` instance is used for a given ruby app, with all threads using it, I think we'll need to take an extra pass at the code while considering "what if this is happening concurrently".

**Comment by remeh** [comment]
> One extra thing that occurred to me is that some of the changes to the `Sender` classes are probably not thread-safe.
> 
> Since the intention seems to be that one `Sender` instance is used for a given ruby app, with all threads using it, I think we'll need to take an extra pass at the code while considering "what if this is happening concurrently".

Yes entirely true, that's exactly what I'm doing right now by adding a mutex, mainly around the message queue. 👍

File: lib/datadog/statsd/udp_connection.rb (hunk 1 above)
Comment (remeh) [reply lib/datadog/statsd/udp_connection.rb]: These magic getter (allocating if nil) are a bad idea overall. I'm removing it and will refactor the rest.

File: lib/datadog/statsd/forwarder.rb [hunk 5]
```diff
@@ -55,7 +55,7 @@ def initialize(
           overflowing_stategy: buffer_overflowing_stategy,
         )
 
-        @sender = single_thread ? SingleThreadSender.new(buffer) : Sender.new(buffer)
+        @sender = single_thread ? SingleThreadSender.new(buffer) : Sender.new(buffer, logger: logger)
```
Comment (ivoanjo) [inline lib/datadog/statsd/forwarder.rb:58]: Minor: I suggest keeping the interface the same between both implementations, including in the constructor here.

One other suggestion to make sure we always use both classes in the same way would be to do: 

```ruby
@sender = (single_thread ? SingleThreadSender : Sender).new(buffer, logger: logger)
```

File: lib/datadog/statsd/sender.rb [hunk 6]
```diff
@@ -2,20 +2,34 @@
...
 
       def flush(sync: false)
-        # don't try to flush if there is no message_queue instantiated
-        return unless message_queue
-
-        message_queue.push(:flush)
+        # don't try to flush if there is no message_queue instantiated or
+        # no companion thread running
+        if !message_queue || !sender_thread.alive?
+          @logger.warn { "Statsd: can't flush: no message queue or sender_thread alive" } if @logger
```
Comment (ivoanjo) [inline lib/datadog/statsd/sender.rb:25]: While this message is accurate, since it's a user-visible warning, if a user sees

```
[WARN] Statsd: can't flush: no message queue or sender_thread alive
```

show up in their logs, I'm not entirely sure they'll understand what the issue is or how they can fix it.

Perhaps we can suggest that they call `#start` in the case where there's no `message_queue`? And in the case of the `sender_thread` not being alive, is there anything we can suggest?

File: lib/datadog/statsd/sender.rb [hunk 7]
```diff
@@ -32,40 +46,56 @@ def rendez_vous
       def add(message)
         raise ArgumentError, 'Start sender first' unless message_queue
 
+        # if the thread does not exist, we assume we are running in a forked process,
+        # empty the message queue and message buffers (these messages belong to
+        # the parent process) and spawn a new companion thread.
+        if !sender_thread.alive?
+          @logger.warn { "Statsd: companion thread dead, re-creating one" } if @logger
```
Comment (ivoanjo) [inline lib/datadog/statsd/sender.rb:53]: Same note as above: if a user sees this in their log, I'm not entirely sure they'll understand the meaning, and if it's anything they can fix or not. I suggest clarifying this from that POV.

File: lib/datadog/statsd/sender.rb [hunk 8]
```diff
@@ -32,40 +46,56 @@ def rendez_vous
...
+          @mx.synchronize {
+            message_queue.close if CLOSEABLE_QUEUES
+            @message_queue = nil
+            message_buffer.reset
+            start
```
Comment (ivoanjo) [inline lib/datadog/statsd/sender.rb:58]: There's a race hiding here. If a process just forked, and started two new threads, and both threads try to report data, the following timeline can occur:

1. Thread A checks line 52, sees that no sender thread is alive. Grabs the lock. (Scheduler switches threads)
2. Thread B checks line 52, sees that no sender thread is alive. Tries to grab the lock, is taken, so it blocks waiting for lock (Scheduler switches threads)
3. Thread A closes previous queue and resets it. Creates a new thread. Releases the lock. Adds message to queue. (Scheduler switches threads)
4. Thread B wakes up, grabs the lock, closes the queue that Thread A created and resets it (data from Thread A) gets lost. Creates another sender thread. etc...

TL;DR After grabbing the lock, we need to check that the world state is as we expected it to be when we went in to grab the lock, as it might've moved in the meanwhile :)

File: lib/datadog/statsd/sender.rb [hunk 9]
```diff
@@ -32,40 +46,56 @@ def rendez_vous
...
+            message_queue = @message_queue
+            message_queue << :close if message_queue
 
-          sender_thread = @sender_thread
-          sender_thread.join if sender_thread && join_worker
+            sender_thread = @sender_thread
+            sender_thread.join if sender_thread && join_worker
+          }
         end
       end
```
Comment (ivoanjo) [inline lib/datadog/statsd/sender.rb:94]: Minor: These two methods are very similar -- would it be worth unifying them and only doing the `if CLOSEABLE_QUEUES` in the one line that changes between them?

File: lib/datadog/statsd/sender.rb [hunk 10]
```diff
@@ -2,20 +2,34 @@
...
+        end
 
-        rendez_vous if sync
+        @mx.synchronize {
+          message_queue.push(:flush)
+          rendez_vous if sync
+        }
       end
 
       def rendez_vous
```
Comment (ivoanjo) [inline lib/datadog/statsd/sender.rb:35]: Since `#rendez_vous` is public, should it also perform similar checks to `#flush` and use the lock as well?

File: lib/datadog/statsd/single_thread_sender.rb [hunk 11]
```diff
@@ -2,12 +2,24 @@
...
+        update_fork_pid
       end
 
       def add(message)
+        # we have just forked, meaning we have messages in the buffer that we should
+        # not send, they belong to the parent process, let's clear the buffer.
+        if forked?
+          @message_buffer.reset
+          update_fork_pid
+        end
```
Comment (ivoanjo) [inline lib/datadog/statsd/single_thread_sender.rb:22]: A similar race to the one I described for the `Sender` can happen here. Two threads call this method at the same time, both of them get `forked?` => `true` and the `@message_buffer` gets reset twice.

Also the interaction with the buffer itself may have issues with concurrency.

File: lib/datadog/statsd/connection.rb [hunk 12]
```diff
@@ -36,6 +36,7 @@ def write(payload)
           retries += 1
           begin
             close
+            connect
```
[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/statsd/connection.rb:39]: After the latest changes, should `def socket` become just a getter, and the `||= connect` removed?

**Review by ivoanjo — COMMENTED** [review]
Sorry for the extra "oh btw concurrency and lots of changes" notes. Feel free to separate changes out to a separate PR if you'd like to get the fork checking in first and separately tackle the concurrency.

File: lib/datadog/statsd/sender.rb (hunk 8 above)
Comment (remeh) [reply lib/datadog/statsd/sender.rb:58]: > If a process just forked, and started two new threads

I don't expect the first process to have to create a thread, but I agree it could happen if for an unknown reason the first process see its thread dying at the same moment. I'll look into this.

File: lib/datadog/statsd/sender.rb (hunk 6 above)
Comment (remeh) [reply lib/datadog/statsd/sender.rb:25]: I'll move these two logs to the `debug` level: they should not happen and are here to help debugging the lib in case of an issue more than anything else.

File: lib/datadog/statsd/connection.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (remeh) [reply lib/datadog/statsd/connection.rb:29]: Fixed in #204

File: lib/datadog/statsd/udp_connection.rb (hunk 1 above)
Comment (remeh) [reply lib/datadog/statsd/udp_connection.rb]: Fixed in #204

**Comment by remeh** [comment]
Addressed by #205
